src/other/broadcasting.jl

```diff
@@ -28,7 +28,12 @@ function Base.copyto!(lazydf::LazyNewColDataFrame, bc::Base.Broadcast.Broadcaste
     if isempty(lazydf.df)
         throw(ArgumentError("creating a column via broadcasting is not allowed on empty data frames"))
     end
-    T = mapreduce(i -> typeof(bc[i]), promote_type, eachindex(bc); init=Union{})
+    if bc isa Base.Broadcast.Broadcasted{<:Base.Broadcast.AbstractArrayStyle{0}} &&
+       bc.f === identity && bc.args isa Tuple{Any} && Base.Broadcast.isflat(bc)
+        T = typeof(bc.args[1][])
```

Comment:
So does this cover all cases where the eltype is known, or would it be useful to have an `if isconcrete(eltype(bc))` branch?

Reply:
`eltype(bc)` is a complex type that does not give information about what will be the type of `bc[idx]`.
 We cover only a simple case when the transformation function is `identity` and we have a scalar that is broadcasted (but I guess this is exactly the important case).

Why in general it is hard to cover? Because you can have something like e.g.:

```
x .= (v -> rand() ? 1 : "a").(1)
```
(I give an extreme case, but in general the point is that `eltype(bc)` is type BEFORE the function applied to the collection is evaluated), the simple case is x.= scalar, as we know that the function is `identity` and we have only one value so we can take it.

Reply:
OK. I wonder whether the compiler is able to compute the element type statically when applying a type-stable function, instead of doing the full `mapreduce` call (which will be quite slow).

Actually, a more standard approach would probably be to allocate a column with the type of the first element, and widen the type if needed like `collect` does. I wonder whether we could reuse some functions from Tables.jl to do that. Or maybe from the grouping code.

Reply:
For example, this seems to work. Maybe there's a better way to avoid the `nothing` hack, though:
```julia
julia> itr = Base.Broadcast.broadcasted(x -> rand() > 0.5 ? 1 : 1.5, 1:100)
Base.Broadcast.Broadcasted(getfield(Main, Symbol("##31#32"))(), (1:100,))

julia> last(Tables.columntable(nothing=>i for i in itr))
100-element Array{Float64,1}:
```

Reply:
Actually we already have `copyto_widen` in grouping.jl. Widening the type of the second argument should be enough, right?

Reply:
We can iteratively `push!` or widen column (this is what Tables.jl essentially does). But I would leave it for later as this does not change API, the performance gains shall not be big (if any - note that now we have ~10% penalty) and the case we are discussing here is rare and can be handled more simply by writing `df[new_col] = vector_value`.

Reply:
> Actually we already have `copyto_widen!` in grouping.jl

I was writing in parallel and did not read this comment. I will have a look at it and report back.

Reply:
This works (I have to minimally tweak `copyto_widen!` for it to work with `bc`):
```
function copyto_widen!(res::AbstractVector{T}, bc::Base.Broadcast.Broadcasted) where T
    @inbounds for i in eachindex(bc)
        val = bc[i]
        S = typeof(val)
        if S <: T || promote_type(S, T) <: T
            res[i] = val
        else
            newres = Tables.allocatecolumn(promote_type(S, T), length(x))
            return copyto_widen!(newres, bc)
        end
    end
    return res
end

function Base.copyto!(lazydf::LazyNewColDataFrame, bc::Base.Broadcast.Broadcasted)
    if isempty(lazydf.df)
        throw(ArgumentError("creating a column via broadcasting is not allowed on empty data frames"))
    end
    if bc isa Base.Broadcast.Broadcasted{<:Base.Broadcast.AbstractArrayStyle{0}} &&
       bc.f === identity && bc.args isa Tuple{Any} && Base.Broadcast.isflat(bc)
        T = typeof(bc.args[1][])
        col = Tables.allocatecolumn(T, nrow(lazydf.df))
        copyto!(col, bc)
    else
        T = typeof(bc[iterate(eachindex(bc))[1]])
        col = copyto_widen!(Tables.allocatecolumn(T, nrow(lazydf.df)), bc)
    end
    lazydf.df[lazydf.col] = col
end
```

It is a bit faster (5% at most) than the current implementation. Do you want to switch to it then?

Reply:
ah - and it will be slower if the "offending" observation with different element type is at the end of the vector (because then it has to "redo" the work twice).
Also - it just threw me `EXCEPTION_ACCESS_VIOLATION` error.
In summary - I would stay with what we have now as it is safe and fast enough.

Reply:
My idea was that you would then be able to drop the branch, since `copyto_widen!` is efficient when the type doesn't change. So that would simplify things. I also don't think it would be slower for the type-unstable case, since the `mapreduce` call will probably be quite slow, in particular since iteration may require evaluating one or several functions for complex cases.

But the crash you mention is worrying. If you can't work around it, can you push a branch somewhere and file an issue in Julia with a reproducer?

Reply:
The `copyto_widen!` approach was very slow when some new type occurred at the end of the vector. Also in general it was a bit slower than `copyto!` in the fast branch.

However, I have just pushed a cleaner approach. What it does it:
* it does normal broadcasting (and has a performance of normal broadcasting) in all most all cases
* except if we have in the end eltype that is categorical in which case one additional allocation is performed (as unfortunately normal broadcasting does not handle it well - but this is a more general problem I am sure you are aware of - if we solve it in general then we also can simplify the code here)

So in short - for now - what I propose should be maximally fast unless we broadcast over a vector of categorical values, in which case one additional copy is performed. I think it is an acceptable compromise.

 And as I have said - if in CategoricalArrays.jl you decide to add a special `Base.Broadcast.materialize` methods for categorical arrays we will be fast in all cases and can remove the unnecessary copy here (if it is not clear what should be added to CategoricalArrays.jl I can expand on it).

Reply:
OK, we need to fix `broadcast` on `CategoricalArray`. Maybe we can get things working by defining `Base.similar(bc::Broadcasted{DefaultArrayStyle{N}}, ::Type{CategoricalString})` and so on. But we should really define something like `allocatecolumn` in Base and have `broadcast` use that, or we'll always suffer from this kind of inconsistency.

Though maybe we'll also need an optimized `materialize` method to speed up special cases, in particular assigning a single `CategoricalValue` to a column (either an existing `CategoricalVector` or a new column). There's also again the issue of propagating levels, which cannot be addressed by just overloading `similar`. What do you think?

Reply:
My thinking is:
### Simple thing - what to do with this PR:
* I have made yet another improvement - now it is fast in all cases except a complex broadcast when we apply a function to a broadcasted `CategoricalVector` that is really unlikely (I have added such complex broadcast to the tests)
* therefore if you are OK with this I would merge this PR

### Difficult thing - what to do in the long run with `CategoricalArray`:

* a half-way solution that will fix this PR in general (even in the complex case) is to make this work correctly:
```
julia> identity.(categorical(["a"]))
1-element Array{CategoricalString{UInt32},1}:
 "a"
```
fixing this is hopefully relatively easy and can be done in CategoricalArrays.jl

* a long term solution is to have `allocatecolumn` or something similar in Base as a general solution so that things like:
```
julia> [v for v in categorical(["a"])]
1-element Array{CategoricalString{UInt32},1}:
 "a"
```
works OK.

Now - I was surprised by this:
```
julia> map(x->x, categorical(["a"]))
1-element CategoricalArray{String,1,UInt32}:
 "a"
```
I did not have time to investigate (probably you know why this works) but maybe the idea used there can be reused (if I replace `CategoricalVector` by broadcasted version of it it does not work though).

Reply:
OK, I think I've found a solution similar to what we did for `map`: https://github.com/JuliaData/CategoricalArrays.jl/pull/193. If we merge and tag a release, do you think we can get rid of the special casing here?

Reply:
Yes - that should fix it.